### PR TITLE
BUG: ``spin gdb``: launch Python directly so that breakpoint is caught

### LINKED ISSUE #24121: BUG: `spin gdb` doesn't hit breakpoints
The new `spin gdb` command added in #24083 starts a gdb session and runs the code under gdb, as expected, but seems to not hit any breakpoints in the executed code.

Compare a python session where I invoke gdb directly, using a python snippet that should hit a breakpoint inside of the `kill` syscall, at least on linux:

```
$ gdb --args $(pyenv which python) -c "import os; import signal; PID = os.getpid(); os.kill(PID, signal.SIGUSR1)"
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/nathan/.pyenv/versions/3.10.9-debug/bin/python...
(gdb) r
Starting program: /home/nathan/.pyenv/versions/3.10.9-debug/bin/python -c import\ os\;\ import\ signal\;\ PID\ =\ os.getpid\(\)\;\ os.kill\(PID,\ signal.SIGUSR1\)
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGUSR1, User defined signal 1.
0x00007ffff764275b in kill () at ../sysdeps/unix/syscall-template.S:120
120	../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) c
Continuing.

Program terminated with signal SIGUSR1, User defined signal 1.
The program no longer exists.
```

However, when I run the same snippet under `spin gdb`, no breakpoint happens:

```
$ python -m spin gdb "import os; import signal; PID = os.getpid(); os.kill(PID, signal.SIGUSR1)"
$ gdb --args python -m spin run python -c 'import os; import signal; PID = os.getpid(); os.kill(PID, signal.SIGUSR1)'
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...
(gdb) r
Starting program: /home/nathan/.pyenv/versions/3.10.9-debug/bin/python -m spin run python -c import\ os\;\ import\ signal\;\ PID\ =\ os.getpid\(\)\;\ os.kill\(PID,\ signal.SIGUSR1\)
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[Detaching after vfork from child process 907000]
[Inferior 1 (process 906997) exited with code 0366]
(gdb) exit


```

ping @stefanv

### PR DESCRIPTION
**Pull request by stefanv**
Closes #24121

### CODE REVIEW

**Comment by ngoldbaum** [comment]
It seems gdb has an option to follow fork to child processes. The following patch seems to work and it retains that it more-or-less uses `spin run` under the hood:

```
diff --git a/.spin/cmds.py b/.spin/cmds.py
index 07e0d6b36..12bb1cbcc 100644
--- a/.spin/cmds.py
+++ b/.spin/cmds.py
@@ -148,8 +148,8 @@ def gdb(python_expr):
 
     """
     util.run(
-        ['gdb', '--args', 'python', '-m', 'spin', 'run',
-         'python', '-P', '-c', python_expr],
+        ['gdb', '-ex', 'set follow-fork-mode child', '--args', 'python', '-m', 'spin', 'run',
+         'python', '-c', python_expr],
         replace=True
     )
```

Fine with the approach in this PR too if you don't think using `spin run` is meaningful here.

**Comment by ngoldbaum** [comment]
Also my Python install seems to complain about the `-P` argument you pass it,

```
Unknown option: -P
usage: python [option] ... [-c cmd | -m mod | file | -] [arg] ...
Try `python -h' for more information.
```

**Comment by stefanv** [comment]
Yeah, `-W` only available in 3.11, so I can go back to "run", let me check.

**Comment by stefanv** [comment]
OK, take another look please. I have to manipulate the sys.path a bit on Python <3.11 to make it work.

Another question about interface: will you always want to run Python code snippets, or should the command allow for arbitrary commands?

**Comment by ngoldbaum** [comment]
For me, I'd want first to check if the argument is a path that exists and then assume that path is a python script, and then don't pass `-c` in the case, otherwise use `-c` and assume the argument is code snippet. I often am debugging longer scripts or have some setup so I can skip breakpoints that would be hit by `import numpy as np`.

**Comment by stefanv** [comment]
I think the interface is now a bit more intuitive.
You can specify code with `-c` like you would for Python, otherwise it handles like `gdb`.

File: .spin/cmds.py [hunk 1]
```diff
@@ -142,16 +142,38 @@ def test(ctx, pytest_args, markexpr, n_jobs, tests, verbose):
 
 
 @click.command()
-@click.argument('python_expr')
-def gdb(python_expr):
+@click.option('--code', '-c', help='Python pogram passed in as a string')
```
Comment (ngoldbaum) [inline .spin/cmds.py:145]:
```suggestion
@click.option('--code', '-c', help='Python program passed in as a string')
```

**Review by ngoldbaum — COMMENTED** [review]
I still think it would be nice to have a mode where you just do `spin gdb my_script.py` and `spin` handles constructing the appropriate arguments to do `gdb --args python my_script.py`. 

My reason for that is admittedly a little niche: with pyenv I need to do things like `gdb --args $(pyenv which python) my_script.py` since the `python` executable in my path is a shim that delegates to the real python executable and gdb needs the real executable. Having a stronger guarantee that the `python` being executed is the same one executing `spin` is also nice.

**Comment by stefanv** [comment]
OK, done.

File: .spin/cmds.py [hunk 2]
```diff
@@ -142,16 +142,46 @@ def test(ctx, pytest_args, markexpr, n_jobs, tests, verbose):
...
+    Or pass arguments to gdb:
+
+    spin gdb -c 'import numpy as np; print(np.__version__)' -- --fullname
+
+    Or run another program, they way you normally would with gdb:
+
+    spin gdb ls
+    spin gdb -- --args ls -al
+
+    You can also run Python progreams:
```
Comment (ngoldbaum) [inline .spin/cmds.py:161]:
```suggestion
    You can also run Python programs:
```

File: .spin/cmds.py [hunk 3]
```diff
@@ -142,16 +142,46 @@ def test(ctx, pytest_args, markexpr, n_jobs, tests, verbose):
...
+        code_prefix = ''
+    else:
+        PYTHON_FLAGS = []
+        code_prefix = 'import sys; sys.path.pop(0); '
+
+    if code:
+        PYTHON_ARGS = ['-c', code_prefix + code]
+        gdb_args += ['--args', sys.executable] + PYTHON_FLAGS + PYTHON_ARGS
+
+    gdb_cmd = ['gdb', '-ex', 'set follow-fork-mode child'] + gdb_args
```
Comment (ngoldbaum) [inline .spin/cmds.py:183]:
```suggestion
    gdb_cmd = ['gdb', '-ex', 'set detach-on-fork on'] + gdb_args
```

When I suggested `follow-fork-mode child` earlier I was testing with a script that didn't import numpy. Actually importing numpy spawns some threads which seems to interfere with `follow-fork-mode child`. It seems we actually want to debug both the child and parent process.

**Review by ngoldbaum — COMMENTED** [review]
Almost there, with the substantive change I suggested `spin gdb test.py` "just works" on my setup for a script that imports numpy, then calls `os.kill()` on itself. :tada:

File: .spin/cmds.py (hunk 2 above)
Comment (stefanv) [reply .spin/cmds.py:161]: Lol

**Comment by stefanv** [comment]
Let's see!

**Review by ngoldbaum — APPROVED** [review]
Let’s let someone else test with a different setup before we merge

**Review by jarrodmillman — APPROVED** [review]

**Comment by stefanv** [comment]
Perhaps @mattip or @seberg?

**Comment by seberg** [comment]
Possible follow-ups I noticed:
* Most other commands rebuild I think, it might be nice to keep it consistent.
* `spin gdb` without anything dosen't do anything useful, it should maybe use `python`?  (`spin python --gdb` would also make sense to do that)

We could extend the help to give tips about setting break-points and start things (maybe even slash down the developer help on it and just say to check `spin gdb --help`?)

Anyway, just noting in case this might spread across many projects really.  I might follow-up with running `lldb` the same way if the system is macos.

**Pull request merged**